File: Swift-version/LeetCode/LeetCode/Solutions/GazeOffer/Solution_21.swift
//
//  Solution_21.swift
//  LeetCode
//
//  Created by 齐旭晨 on 2021/3/30.
//

import Foundation

class Solution_21: Solution{
    func main() {
        print(exchange([1,2,3,4]))
    }
    
    func exchange(_ nums: [Int]) -> [Int] {
        var nums = nums
        var left = nums.startIndex
        var right = nums.endIndex - 1
        while(left < right){
            while(left < right && nums[left] % 2 == 1){
                left = nums.index(after: left)
            }
            while(left < right && nums[right] % 2 == 0){
                right = nums.index(before: right)
            }
            nums.swapAt(left, right)
        }
        return nums
    }
}

File: Java-version/LeetCode/src/com/jooc/NewCoder/Solution_21.java
package com.jooc.NewCoder;

import com.jooc.Solution;

public class Solution_21 implements Solution {
    @Override
    public void main() {
        System.out.println(minNumberInRotateArray(new int[]{1,0,1,1,1}));
    }

    public int minNumberInRotateArray(int[] array) {
        int left = 0, right = array.length - 1;

        while (left < right) {
            int mid = (left + right) / 2;
            if (array[mid] == array[right])
                right--;
            else if (array[mid] > array[right])
                left = mid + 1;
            else
                right = mid;
        }

        return array[left];
    }
}
